### PR TITLE
Fix #7052: Transform annotations in ReifyQuotes

### LINKED ISSUE #7052: ReifyQuotes is not transforming the annotations
## minimized code

```Scala
import scala.quoted._
class Test {
  def foo(str: Expr[String]) given QuoteContext = '{
    @deprecated($str, "")
    def a = ???
  }
}
```

<details>
<summary>Stack trace</summary>

```scala
java.lang.AssertionError: assertion failed: unresolved symbols: value str(line 2) when pickling tests/run-with-compiler/foo.scala while compiling tests/run-with-compiler/foo.scala
Exception in thread "main" java.lang.AssertionError: assertion failed: unresolved symbols: value str(line 2) when pickling tests/run-with-compiler/foo.scala
        at dotty.DottyPredef$.assertFail(DottyPredef.scala:16)
        at dotty.tools.dotc.core.tasty.TreePickler.pickle(TreePickler.scala:695)
        at dotty.tools.dotc.core.quoted.PickledQuotes$.pickle(PickledQuotes.scala:90)
        at dotty.tools.dotc.core.quoted.PickledQuotes$.pickleQuote(PickledQuotes.scala:32)
        at dotty.tools.dotc.transform.ReifyQuotes$QuoteReifier.pickleAsTasty$1(ReifyQuotes.scala:237)
        at dotty.tools.dotc.transform.ReifyQuotes$QuoteReifier.pickledQuote(ReifyQuotes.scala:249)
        at dotty.tools.dotc.transform.ReifyQuotes$QuoteReifier.transformQuotation(ReifyQuotes.scala:195)
        at dotty.tools.dotc.transform.TreeMapWithStages.transform(TreeMapWithStages.scala:94)
        at dotty.tools.dotc.transform.ReifyQuotes$QuoteReifier.transform(ReifyQuotes.scala:410)
        at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1256)
        at dotty.tools.dotc.ast.TreeMapWithImplicits.transform(TreeMapWithImplicits.scala:118)
        at dotty.tools.dotc.transform.TreeMapWithStages.mapOverTree$1(TreeMapWithStages.scala:77)
        at dotty.tools.dotc.transform.TreeMapWithStages.transform(TreeMapWithStages.scala:127)
        at dotty.tools.dotc.transform.ReifyQuotes$QuoteReifier.transform(ReifyQuotes.scala:410)
        at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1262)
        at dotty.tools.dotc.ast.TreeMapWithImplicits.transform(TreeMapWithImplicits.scala:118)
        at dotty.tools.dotc.transform.TreeMapWithStages.mapOverTree$1(TreeMapWithStages.scala:77)
        at dotty.tools.dotc.transform.TreeMapWithStages.transform(TreeMapWithStages.scala:127)
        at dotty.tools.dotc.transform.ReifyQuotes$QuoteReifier.transform(ReifyQuotes.scala:410)
        at dotty.tools.dotc.ast.TreeMapWithImplicits.transform(TreeMapWithImplicits.scala:98)
        at dotty.tools.dotc.transform.TreeMapWithStages.mapOverTree$1(TreeMapWithStages.scala:77)
        at dotty.tools.dotc.transform.TreeMapWithStages.transform(TreeMapWithStages.scala:127)
        at dotty.tools.dotc.transform.ReifyQuotes$QuoteReifier.transform(ReifyQuotes.scala:410)
        at dotty.tools.dotc.ast.TreeMapWithImplicits.traverse$1(TreeMapWithImplicits.scala:52)
        at dotty.tools.dotc.ast.TreeMapWithImplicits.transformStats(TreeMapWithImplicits.scala:59)
        at dotty.tools.dotc.ast.TreeMapWithImplicits.transform(TreeMapWithImplicits.scala:109)
        at dotty.tools.dotc.transform.TreeMapWithStages.mapOverTree$1(TreeMapWithStages.scala:77)
        at dotty.tools.dotc.transform.TreeMapWithStages.transform(TreeMapWithStages.scala:127)
        at dotty.tools.dotc.transform.ReifyQuotes$QuoteReifier.transform(ReifyQuotes.scala:410)
        at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1332)
        at dotty.tools.dotc.ast.TreeMapWithImplicits.transform(TreeMapWithImplicits.scala:102)
        at dotty.tools.dotc.transform.TreeMapWithStages.mapOverTree$1(TreeMapWithStages.scala:77)
        at dotty.tools.dotc.transform.TreeMapWithStages.transform(TreeMapWithStages.scala:127)
        at dotty.tools.dotc.transform.ReifyQuotes$QuoteReifier.transform(ReifyQuotes.scala:410)
        at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform$$anonfun$2(Trees.scala:1352)
        at scala.collection.immutable.List.mapConserve(List.scala:179)
        at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1352)
        at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformStats(Trees.scala:1350)
        at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1338)
        at dotty.tools.dotc.ast.TreeMapWithImplicits.transform(TreeMapWithImplicits.scala:102)
        at dotty.tools.dotc.transform.TreeMapWithStages.mapOverTree$1(TreeMapWithStages.scala:77)
        at dotty.tools.dotc.transform.TreeMapWithStages.transform(TreeMapWithStages.scala:127)
        at dotty.tools.dotc.transform.ReifyQuotes$QuoteReifier.transform(ReifyQuotes.scala:410)
        at dotty.tools.dotc.transform.ReifyQuotes$$anon$1.transform(ReifyQuotes.scala:94)
        at dotty.tools.dotc.transform.MacroTransform.run(MacroTransform.scala:21)
        at dotty.tools.dotc.transform.ReifyQuotes.run(ReifyQuotes.scala:90)
        at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:316)
        at scala.collection.immutable.List.map(List.scala:286)
        at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:318)
        at dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:158)
        at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:36)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:33)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:198)
        at dotty.tools.dotc.Run.runPhases$5(Run.scala:170)
        at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:178)
        at dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
        at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:65)
        at dotty.tools.dotc.Run.compileUnits(Run.scala:185)
        at dotty.tools.dotc.Run.compileSources(Run.scala:120)
        at dotty.tools.dotc.Run.compile(Run.scala:104)
        at dotty.tools.dotc.Driver.doCompile(Driver.scala:35)
        at dotty.tools.dotc.Driver.process(Driver.scala:180)
        at dotty.tools.dotc.Driver.process(Driver.scala:149)
        at dotty.tools.dotc.Driver.process(Driver.scala:161)
        at dotty.tools.dotc.Driver.main(Driver.scala:188)
        at dotty.tools.dotc.Main.main(Main.scala)```
</details>

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

**Comment by nicolasstucki** [comment]
@liufengyun the two points where addressed

File: compiler/src/dotty/tools/dotc/ast/TreeMapWithImplicits.scala [hunk 1]
```diff
@@ -62,7 +62,7 @@ class TreeMapWithImplicits extends tpd.TreeMap {
   private def nestedScopeCtx(defs: List[Tree])(implicit ctx: Context): Context = {
     val nestedCtx = ctx.fresh.setNewScope
     defs foreach {
-      case d: DefTree => nestedCtx.enter(d.symbol)
+      case d: DefTree if d.symbol.isOneOf(GivenOrImplicit) => nestedCtx.enter(d.symbol)
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/ast/TreeMapWithImplicits.scala]: This change seems to break the originally intended semantics?

File: compiler/src/dotty/tools/dotc/ast/TreeMapWithImplicits.scala [hunk 2]
```diff
@@ -73,7 +73,7 @@ class TreeMapWithImplicits extends tpd.TreeMap {
     new TreeTraverser {
       def traverse(tree: Tree)(implicit ctx: Context): Unit = {
         tree match {
-          case d: DefTree => nestedCtx.enter(d.symbol)
+          case d: DefTree if d.symbol.isOneOf(GivenOrImplicit) => nestedCtx.enter(d.symbol)
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/ast/TreeMapWithImplicits.scala]: Same question as above.

File: compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala [hunk 3]
```diff
@@ -405,6 +405,14 @@ class ReifyQuotes extends MacroTransform {
...
             cpy.DefDef(tree)(rhs = defaultValue(tree.rhs.tpe))
 
+          case tree: DefTree if level >= 1 =>
+            val newAnnotations = tree.symbol.annotations.mapconserve { annot =>
+              val newAnnotTree = transform(annot.tree) given ctx.withOwner(tree.symbol)
+              if (annot.tree == newAnnotTree) annot
+              else ConcreteAnnotation(newAnnotTree)
+            }
+            tree.symbol.annotations = newAnnotations
+            super.transform(tree)
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala]: Why we need `level > 1`?

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala]: For code like this
```scala
import scala.quoted._
class Test {
  def foo(str: String) given QuoteContext = '{
   '{
     @deprecated(str, "") // error
      def bar = ???
    }
  }
}
```
I will add a test case

File: compiler/src/dotty/tools/dotc/ast/TreeMapWithImplicits.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/ast/TreeMapWithImplicits.scala]: It is supposed to add the missing implicits into the scope. But adding other definitions interacts badly with on of the Yckeck while transforming an annotation containing a definition.

File: compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala (hunk 3 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala]: Added

Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala]: I'm wondering if the following code is possible and if it's handled by the compiler.

```Scala
import scala.quoted._

class annot[T: Type](x: T)

class Test()(given qtx: QuoteContext) {
  @annot(4)
  def foo(str: String) = ()
}
```

**Comment by liufengyun** [comment]
@nicolasstucki The printing of the following program shows that the `'{4}` in the annotation is not reified during compilation:

```Scala
import scala.quoted._

class annot[T: Type](x: Expr[T]) extends scala.annotation.Annotation

class Test()(implicit qtx: QuoteContext) {
  @annot('{4})
  def foo(str: String) = '{4}
}
```

I'm not sure if it's a bug or not, as annotations don't have runtime semantics.

**Comment by nicolasstucki** [comment]
We should probably disallow quotes inside annotations.

**Comment by liufengyun** [comment]
@nicolasstucki Do you prefer to handle the case in this PR, or in a later PR?

**Comment by nicolasstucki** [comment]
Maybe let's open a separate issue.

**Comment by liufengyun** [comment]
The issue is created: https://github.com/lampepfl/dotty/issues/7123

**Pull request merged**